codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/parsers/JsonSerializerGenerator.kt

```diff
@@ -5,75 +5,156 @@
 import software.amazon.smithy.model.shapes.MemberShape
+import software.amazon.smithy.model.shapes.NumberShape
 import software.amazon.smithy.model.shapes.OperationShape
+import software.amazon.smithy.model.shapes.Shape
+import software.amazon.smithy.model.shapes.StringShape
 import software.amazon.smithy.model.shapes.StructureShape
+import software.amazon.smithy.model.shapes.TimestampShape
+import software.amazon.smithy.model.shapes.UnionShape
+import software.amazon.smithy.model.traits.EnumTrait
+import software.amazon.smithy.model.traits.JsonNameTrait
+import software.amazon.smithy.model.traits.TimestampFormatTrait.Format.EPOCH_SECONDS
+import software.amazon.smithy.rust.codegen.rustlang.CargoDependency
+import software.amazon.smithy.rust.codegen.rustlang.RustType
+import software.amazon.smithy.rust.codegen.rustlang.RustWriter
+import software.amazon.smithy.rust.codegen.rustlang.asType
+import software.amazon.smithy.rust.codegen.rustlang.rust
 import software.amazon.smithy.rust.codegen.rustlang.rustBlock
 import software.amazon.smithy.rust.codegen.rustlang.rustBlockTemplate
+import software.amazon.smithy.rust.codegen.rustlang.rustInline
 import software.amazon.smithy.rust.codegen.rustlang.rustTemplate
 import software.amazon.smithy.rust.codegen.rustlang.withBlock
 import software.amazon.smithy.rust.codegen.smithy.RuntimeType
+import software.amazon.smithy.rust.codegen.smithy.RustSymbolProvider
 import software.amazon.smithy.rust.codegen.smithy.generators.ProtocolConfig
+import software.amazon.smithy.rust.codegen.smithy.isOptional
+import software.amazon.smithy.rust.codegen.smithy.rustType
 import software.amazon.smithy.rust.codegen.smithy.traits.SyntheticInputTrait
+import software.amazon.smithy.rust.codegen.util.dq
 import software.amazon.smithy.rust.codegen.util.expectTrait
+import software.amazon.smithy.rust.codegen.util.getTrait
+import software.amazon.smithy.rust.codegen.util.hasTrait
 import software.amazon.smithy.rust.codegen.util.inputShape
+import software.amazon.smithy.rust.codegen.util.toPascalCase
 import software.amazon.smithy.rust.codegen.util.toSnakeCase
 
+private data class SimpleContext<T : Shape>(
+    /** Name of the JsonObjectWriter or JsonArrayWriter */
+    val writerName: String,
+    val localName: String,
+    val shape: T,
+)
+
+private data class StructContext(
+    /** Name of the JsonObjectWriter */
+    val objectName: String,
+    val localName: String,
+    val shape: StructureShape,
+    val symbolProvider: RustSymbolProvider,
+) {
+    fun member(member: MemberShape): MemberContext =
+        MemberContext(objectName, "$localName.${symbolProvider.toMemberName(member)}", member, structMember = true)
+}
+
+private data class MemberContext(
+    /** Name of the JsonObjectWriter or JsonArrayWriter */
+    val writerName: String,
+    val valueExpression: String,
+    val shape: MemberShape,
+    /** Whether we're working with a JsonObjectWriter (true) or JsonArrayWriter (false) */
+    val structMember: Boolean,
```

Comment:
this was a little confusing when I was reading the code, might make a sealed class for this which will make the usage:
```kotlin
when(parent) {
  is Json.Object -> ...
  is Json.Array -> ...
}
```

which I think will be much clearer